### PR TITLE
west_commands: runners: openocd: clear DHCSR after flashing completes

### LINKED ISSUE #32984: West: openocd runner: Don't let debug mode on by default
**Describe the bug**
When flashing using openocd runner, the cortex-m processor is left in debug mode. It can, then, only be released after a power-on reset (cold reset).

This conflicts with the use of https://github.com/zephyrproject-rtos/zephyr/pull/31481 (Null pointer exception detection via DWT), and prevent its use by default in tests.

~~For this reason, Null pointer exception detection was disabled by default on all STM32 platforms, but it would be great to be able to re-instantiate it.~~

Note: This potentially affect other runners as well (~~pyocd~~, for instance).

**To Reproduce**
1. Revert https://github.com/zephyrproject-rtos/zephyr/pull/32983
2. Run any zephyr test on a STM32 target using openocd runner for flashing

**Expected behavior**
Don't prevent use of Null pointer exception detection when using openocd runner

**Impact**
Prevent use of Null pointer exception detection when using openocd runner in zephyr tests

**Logs and console output**
If applicable, add console logs or other types of debug information
e.g Wireshark capture or Logic analyzer capture (upload in zip archive).
copy-and-paste text and put a code fence (\`\`\`) before and after, to help
explain the issue. (if unable to obtain text log, add a screenshot)

**Environment (please complete the following information):**
 - OS: (e.g. Linux, MacOS, Windows)
 - Toolchain (e.g Zephyr SDK, ...)
 - Commit SHA or Version used

**Additional context**

### PR DESCRIPTION
**Pull request by ioannisg**
Fixes #32984 

openOCD does not clear DHCSR.DEBUGEN bit after flashing
completes, effectively leaving the Cortex-M processor in
debug mode. Manually clean DHCSR becore exiting the flash
operation, to work around the problem.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

### CODE REVIEW

**Comment by ioannisg** [comment]
This is just a trial for now - do not merge, since the fix should only be applicable for Cortex-M.
@erwango could you check if this would work for openOCD-flashed targets?

File: scripts/west_commands/runners/openocd.py
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/openocd.py:171]: Thanks for this. No comments on the register writes themselves, but thoughts on getting this out of draft:

To abstract this properly, I think we need to either:

1. move these into one of the existing command line hooks (e.g. the one that ends up in `post_verify_cmd`), or
2. create a new command line hook

After that, we can either:

1. make a `boards/common/openocd-cortex-m.board.cmake` file that sets this up appropriately, and include it from all the right board.cmake files (not forgetting to update `openocd-nrf5.board.cmake` too), or
2. modify openocd.board.cmake to check Kconfig variables for Cortex-M and do it automatically the way it already looks at `CONFIG_X86` and `CONFIG_ARC` to decide if it uses `load_image` or `flash write_image erase`

Based on the way you've written this PR, it looks like you'll need to add a new command line option since this doesn't appear next to any of the existing `_cmd` lists here or below. I think modifying `openocd.board.cmake` instead of creating a new file will be better and easier.

Comment (mbolivar-nordic) [reply scripts/west_commands/runners/openocd.py:171]: I'm also kind of surprised nobody noticed this yet. Is there a standard openocd command for this that doesn't require touching the register directly?

Comment (galak) [reply scripts/west_commands/runners/openocd.py:171]: > I'm also kind of surprised nobody noticed this yet. Is there a standard openocd command for this that doesn't require touching the register directly?

Not that we are aware of.  There's nothing in the code that is obvious.

Comment (ioannisg) [reply scripts/west_commands/runners/openocd.py:171]: I sent a patch request upstream: http://openocd.zylin.com/#/c/6093/1 to change the default mode.

File: scripts/west_commands/runners/openocd.py
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/openocd.py:182]: This name is confusing mechanism with policy. Please rename it to `post_reset_cmd` instead.

File: boards/common/openocd.board.cmake
Comment (mbolivar-nordic) [inline boards/common/openocd.board.cmake:19]: Can you add some more comments here, since this looks "weird"? Maybe just linking to the openocd PR you opened would be good enough.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Thanks @ioannisg . I read the discussion in upstream openocd and it looks like you really found something new here. A couple of minor requests and then I am +1 from the runner perspective.

However, I want to confirm with @erwango and other openocd users that we really want this to be the default behavior before merging, since the openocd maintainers named some use cases where they didn't think it made a sensible default.

**Comment by ioannisg** [comment]
> However, I want to confirm with @erwango and other openocd users that we really want this to be the default behavior before merging, since the openocd maintainers named some use cases where they didn't think it made a sensible default.

Fine. I 've, also added the HW block label, so this is tested before it (ever) gets merged.

My understanding is that "west flash" shall only program the binary and reset the core to a state as close as possible to the (cold) reset state. Since the internal runner implementation does not guarantee this, we are left with the option to do this in the runner.

File: scripts/west_commands/runners/openocd.py
Comment (ioannisg) [reply scripts/west_commands/runners/openocd.py:182]: Yeap, that's fine.

File: boards/common/openocd.board.cmake
Comment (ioannisg) [reply boards/common/openocd.board.cmake:19]: I'll comment properly.

**Comment by erwango** [comment]
@ioannisg is this PR still required if http://openocd.zylin.com/#/c/6093/1 gets merged ?

**Comment by erwango** [comment]
One consideration on current behavior is that having DHCSR set after flash is that it prevents target entering sleep states when testing PM feature.
In case DHCSR is cleared and PM sleep state is entered, flashing using openocd is not possible anymore and another flashing method shoud be used.
So, having DHCSR set after flash:
- is bad for measurements... BUT
- allows PM testing in CI with existing flashing methods.

File: scripts/west_commands/runners/pyocd.py
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/pyocd.py:166]: This is not correct usage. In Python, `foo is bar` and `foo is not bar` basically work like pointer comparison in C.

Example:

```py
>>> [] is []
False
```

This should be `if self.post_flash_cmd:` instead, because an empty list is boolean false.

File: boards/common/pyocd.board.cmake
Comment (mbolivar-nordic) [inline boards/common/pyocd.board.cmake:17]: This is reading from an uninitialized variable on non-Cortex-M

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]
Nacking until we have an alternative viable option to allow Power Management testing in CI runs.

As previously stated, current openocd behavior regarding debug enable bit allows to flash target after sleep mode code has been exercised, while the proposed change prevents this.

**Comment by galak** [comment]
@ioannisg any updates here?  We should set the default of the feature to disabled until we resolve a number of these issues.

**Pull request closed without merging**
